core/parser.mly

```diff
@@ -443,6 +446,15 @@ tlfunbinding:
 | OP OPERATOR pattern perhaps_location block                   { ((dl_unl, false), $2, [[$3]], $4, $5)          }
 | OP pattern OPERATOR perhaps_location block                   { ((dl_unl, false), $3, [[$2]], $4, $5)          }
 
+match_tlfunbinding:
+| fun_kind VARIABLE arg_lists perhaps_location match_body      { ($1, $2, $3, $4, $5)                }
+
+match_body:
+| MATCH LBRACE match_cases* RBRACE                             { $3 }
+
+match_cases:
+| VBAR pattern RARROW block_contents                           { ($2, block ~ppos:$loc $4) }
```

Comment:
This production matches a single case, not cases.  (so call the nonterminal `match_case`)